assembly/src/ast/tests.rs

```diff
@@ -1171,3 +1171,114 @@ fn assert_parsing_line_unexpected_token() {
         r#" help: expected "@", or "begin", or "const", or "export", or "proc", or "use", or end of file, or doc comment"#
     );
 }
+
+#[test]
+fn test_roundtrip_formatting() {
+    let source = "\
+#! module doc
+#!
+#! with spaces
+
+#! constant doc
+#!
+#! with spaces
+const.DEFAULT_CONST=100
+
+#! Perform `a + b`, `n` times
+#!
+#! with spaces
+proc.add_n_times # [n, b, a]
+    dup.0
+    push.0
+    u32gt
+    if.true
+        push.0.1
+        while.true  # [total, n, b, a]
+            dup.3 dup.3
+            u32wrapping_add3 # [total', n, b, a]
+            swap.1
+            push.1
+            u32overflowing_sub  # [overflowed, n - 1, total', b, a]
+            swap.1 movdn.3      # [overflowed, total', n', b, a]
+            push.0              # [0, overflowed, total, n', total', b, a]
+            dup.1               # [overflowed, 0, overflowed, total', n', b, a]
+            cdrop               # [continue, total', n', b, a]
+        end
+        movdn.3
+        drop drop drop
+    else
+        u32wrapping_add
+    end
+end
+
+begin
+    push.1.1.DEFAULT_CONST
+    exec.add_n_times
+    push.20
+    assert_eq
+
+    trace.DEFAULT_CONST
+end
+";
+
+    let context = TestContext::default();
+    let source = source_file!(&context, source);
+
+    let module = Module::parse(
+        LibraryPath::new_from_components(LibraryNamespace::Exec, []),
+        ModuleKind::Executable,
+        source,
+    )
+    .unwrap_or_else(|err| panic!("{err}"));
+
+    let formatted = module.to_string();
+    let expected = "\
+#! module doc
+#!
+#! with spaces
+
+#! Perform `a + b`, `n` times
+#!
+#! with spaces
+proc.add_n_times
```

Comment:
Similarly here, comments on the same line as code got removed

Reply:
Same as previous comment.